scarlet/source.py

```diff
@@ -395,6 +396,96 @@ def __init__(self, frame, sky_coord, observations):
     def _psf_wrapper(self, *parameters):
         return self.frame.psf.__call__(*parameters, bbox=self.bbox)[0]
 
+class WaveletSource(FunctionComponent):
+    """Source intialized with a single pixel
+
+    Point sources are initialized with the SED of the center pixel,
+    and the morphology taken from `frame.psfs`, centered at `sky_coord`.
+    """
+
+    def __init__(
+        self,
+        frame,
+        sky_coord,
+        observations,
+        noise,
+        obs_idx=0,
+        thresh=1.0,
+        shifting=False,
+    ):
+        """Extended source intialized to match a set of observations
+
+        Parameters
+        ----------
+        frame: `~scarlet.Frame`
+            The frame of the model
+        sky_coord: tuple
+            Center of the source
+        observations: instance or list of `~scarlet.observation.Observation`
+            Observation(s) to initialize this source.
+        obs_idx: int
+            Index of the observation in `observations` to
+            initialize the morphology.
+        thresh: `float`
+            Multiple of the backround RMS used as a
+            flux cutoff for morphology initialization.
+        shifting: `bool`
+            Whether or not a subpixel shift is added as optimization parameter
+        """
+        center = np.array(frame.get_pixel(sky_coord), dtype="float")
+        self.pixel_center = tuple(np.round(center).astype("int"))
+
+        # initialize SED from sky_coord
+        try:
+            iter(observations)
+        except TypeError:
+            observations = [observations]
+
+        # initialize from observation
+        sed, morph, bbox = init_extended_source(
+            sky_coord,
+            frame,
+            observations,
+            obs_idx=obs_idx,
+            thresh=thresh,
+            symmetric=True,
+            monotonic=True,
+        )
+
+        sed = Parameter(
+            sed,
+            name="sed",
+            step=partial(relative_step, factor=1e-2),
+            constraint=PositivityConstraint(),
+        )
+        # Threshold in units of noise
+        thresh *= np.sum(sed*noise)
+
+        # Starlet transform
+        transform = Starlet(morph)
+        morph = transform.starlet
+        # wavelet-scale norm
+        wavelet_norm = transform.norm
+        #One threshold per wavelet scale: thresh*norm
+        thresh_array = np.zeros(morph.shape) + thresh
+        thresh_array[-3:] = thresh_array[-3:] * np.array([wavelet_norm])[..., np.newaxis, np.newaxis]
+        # We don't threshold the last scale
+        thresh_array[...,-1,:,:] = np.inf
+
+        morph_constraint = L0Constraint(thresh_array)
+        morph = Parameter(morph, name="morph", step=1e-2, constraint=morph_constraint)
```

Comment:
What's the amplitude of the morph coefficients? I wonder if `step=1e-2` is appropriate

Reply:
I think it's fine. Coefficients are fractions of the pixel values, fraction ranging from 0.8 to 0.01 for images with a 100 ish pixels.

Reply:
That's quite a large variation. Right now the small coefficients have step size of order unity per iteration. Have you tried with a relative step size? https://github.com/pmelchior/scarlet/blob/master/scarlet/parameter.py#L109
By default this is 10% of the mean along axis=0. This could easily be modified, but I found that much more suitable unless there's a known absolute scale of the parameter

Reply:
I'll try that, but there aree other issues in my implementation I think at the moment that have a larger impact.

Reply:
It returns a nan. I suspect a division by zero.

Reply:
Yes, I'm also getting weird results from the thresholded coefficients. Something isn't right, I think, fixing the step sizes just masks a more profound problem.